presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUtil.java

```diff
@@ -141,6 +143,15 @@ public static Optional<Long> resolveSnapshotIdByName(Table table, IcebergTableNa
         return tryGetCurrentSnapshot(table).map(Snapshot::snapshotId);
     }
 
+    public static long getSnapshotIdAsOfTime(Table table, long millisUtc)
+    {
+        return table.history().stream()
+                .filter(logEntry -> logEntry.timestampMillis() <= millisUtc)
+                .max(comparing(HistoryEntry::timestampMillis))
+                .orElseThrow(() -> new PrestoException(INVALID_ARGUMENTS, format("No history found based on timestamp for table %s", table.name())))
```

Comment:
I would use a new error code in `IcebergErrorCode`, perhaps `ICEBERG_INVALID_TABLE_TIMESTAMP`